code/modules/shuttle/manipulator.dm

```diff
@@ -9,5 +9,214 @@
 	icon = 'icons/obj/machines/dominator.dmi'
 	icon_state = "dominator-blue"
 
+	// UI state variables
+	var/datum/map_template/shuttle/selected
+	var/obj/docking_port/mobile/existing_shuttle
+
+	var/obj/docking_port/mobile/preview_shuttle
+	var/preview_shuttle_id
+
 /obj/machinery/shuttle_manipulator/process()
 	return
+
+/obj/machinery/shuttle_manipulator/ui_interact(mob/user, ui_key = "main", \
+	datum/tgui/ui = null, force_open = 0, datum/tgui/master_ui = null, \
+	datum/ui_state/state = admin_state)
+
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "shuttle_manipulator", name, 800, 600, \
+			master_ui, state)
+		ui.open()
+
+/proc/shuttlemode2str(mode)
+	switch(mode)
+		if(SHUTTLE_IDLE)
+			. = "idle"
+		if(SHUTTLE_RECALL)
+			. = "recalled"
+		if(SHUTTLE_CALL)
+			. = "called"
+		if(SHUTTLE_DOCKED)
+			. = "docked"
+		if(SHUTTLE_STRANDED)
+			. = "stranded"
+		if(SHUTTLE_ENDGAME)
+			. = "endgame"
+	if(!.)
+		throw EXCEPTION("shuttlemode2str(): invalid mode [mode]")
+
+/obj/machinery/shuttle_manipulator/ui_data(mob/user)
+	var/list/data = list()
+	data["tabs"] = list("Status", "Templates", "Modification")
+
+	// Templates panel
+	data["templates"] = list()
+	var/list/templates = data["templates"]
+	data["templates_tabs"] = list()
+	data["selected"] = list()
+
+	for(var/shuttle_id in shuttle_templates)
+		var/datum/map_template/shuttle/S = shuttle_templates[shuttle_id]
+
+		if(!templates[S.port_id])
+			data["templates_tabs"] += S.port_id
+			templates[S.port_id] = list(
+				"port_id" = S.port_id,
+				"templates" = list())
+
+		var/list/L = list()
+		L["name"] = S.name
+		L["shuttle_id"] = S.shuttle_id
+		L["port_id"] = S.port_id
+		L["description"] = S.description
+		L["admin_notes"] = S.admin_notes
+
+		if(selected && (selected.shuttle_id == S.shuttle_id))
```

Comment:
I'd prefer that this be a flag of some sort... Maybe have selected be the name? Sending data down the wire twice when it can be derived is wasteful.

(a computed variable could easily be used)